multisig/src/lib.rs

```diff
@@ -192,16 +185,29 @@ impl MultiSigContract {
                                 .map(|x| x.into())
                                 .unwrap_or(DEFAULT_ALLOWANCE),
                             permission.receiver_id,
-                            permission
-                                .method_names.join(",").into_bytes(),
+                            permission.method_names.join(",").into_bytes(),
                         )
                     } else {
                         // wallet UI should warn user if receiver_id == env::current_account_id(), adding FAK will render multisig useless
                         promise.add_full_access_key(public_key.into())
                     }
                 }
                 MultiSigRequestAction::DeleteKey { public_key } => {
-                    promise.delete_key(public_key.into())
+                    self.assert_self_request(receiver_id.clone());
+                    let pk: PublicKey = public_key.into();
+                    // delete outstanding requests by public_key
+                    let request_ids: Vec<u32> = self
+                        .requests
+                        .iter()
+                        .filter(|(_k, r)| r.signer_pk == pk)
+                        .map(|(k, _r)| k)
+                        .collect();
+                    for request_id in request_ids {
+                        self.remove_request(request_id);
```

Comment:
It's somewhat inefficient to call `remove_request` because it updates `num_requests_pk`, and this requires ton of read/writes. I think it's better to copy only the relevant part there, because you'll nuke the `num_requests_pk` for this entry at the end anyway